### PR TITLE
SD-107: Dual Axes for Column, Bar and Line chart

### PR DESCRIPTION
**Pull request by naibinh**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1892/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1380/ : SUCCESS in 2m 30s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1906/ : SUCCESS in 22s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1391/ : SUCCESS in 4m 48s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1908/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1393/ : SUCCESS in 3m 37s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/781/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1909/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/417/ : UNSTABLE in 18m 17s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1911/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1396/ : SUCCESS in 2m 09s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 1]
```diff
@@ -1139,6 +1149,9 @@ function getYAxes(
     const hasMoreThanOneMeasure = measureGroupItems.length > 1;
     const noPrimaryMeasures = isBucketEmpty(mdObject, MEASURES);
 
+    const { measures : secondMeasures = [] as string[] } =
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: secondaryAxisMeasures would be better name

File: stories/core_components/BarChart.tsx [hunk 2]
```diff
@@ -323,4 +324,42 @@ storiesOf('Core components/BarChart', module)
...
+    ))
+    .add('dual axes with two bottom measures, one top measure, one attribute', () => (
+        screenshotWrap(
+            <div style={wrapperStyle}>
+                <BarChart
+                    projectId="storybook"
+                    measures={[MEASURE_1, MEASURE_2, MEASURE_3]}
+                    config={{
+                        secondary_xaxis: {
+                            measures: [MEASURE_3.measure.localIdentifier]
```
Comment (ivan-nejezchleb) [inline stories/core_components/BarChart.tsx]: try to put here M2 identifier and observe result

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 3]
```diff
@@ -1181,6 +1194,32 @@ function getYAxes(
...
+    } else if (isOneOfTypes(type, supportedDualAxesChartTypes) &&
+                !isEmpty(measureGroupItems) &&
+                !isEmpty(secondMeasures)) {
+
+        const { measuresInFirstAxis, measuresInSecondAxis } = separateAxes(secondMeasures, measureGroup);
+
+        let firstAxis = createYAxisItem(measuresInFirstAxis, false);
+        let secondAxis = createYAxisItem(measuresInSecondAxis, true);
+
+        if (!isEmpty(firstAxis)) {
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: I see that `isEmpty` can be used for object, array and string.

File: src/components/visualizations/chart/highcharts/helpers.ts [hunk 4]
```diff
@@ -238,9 +251,9 @@ function getMinFromPositiveNegativeStacks(data: number[]): number {
     }, Number.MAX_SAFE_INTEGER);
 }
 
-export function shouldStartOnTick(chartOptions: any): boolean {
-    const min = parseFloat(get(chartOptions, 'yAxisProps.min', ''));
-    const max = parseFloat(get(chartOptions, 'yAxisProps.max', ''));
+export function shouldStartOnTick(chartOptions: any, axisPropsKey = 'yAxisProps'): boolean {
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/helpers.ts]: I see that `shouldStartOnTick` is called in unit test as well.
So default value will avoid failed tests.

File: src/components/visualizations/chart/legend/LegendList.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+export const LegendSeparator = () => (<div className="legend-separator"/>);
+
+export default function LegendList(props: any) {
+    const { series, chartType, onItemClick, width } = props;
+    return (
+        series.map((item: any, index: number) => {
+            const { type, text } = item;
+            if (type === LEGEND_AXIS_INDICATOR) {
+                return <LegendAxisIndicator text={text} key={index} width={width}/>;
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/legend/LegendList.tsx]: I don't think we need `()` to wrap one line JSX.
But it's still a good practice to follow though.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1918/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1404/ : FAILURE in 5m 05s

File: src/components/visualizations/chart/chartOptionsBuilder.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: it is true, but you can imagine how complex this check is inside and here you exactly know what to expect from 'createYAxisItem' function

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1919/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1405/ : SUCCESS in 2m 32s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1920/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1406/ : SUCCESS in 2m 23s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 6]
```diff
@@ -1197,6 +1236,53 @@ function getYAxes(
     return yAxes;
 }
 
+function separateAxes(secondMeasures: string[], measureGroup: Execution.IMeasureGroupHeader['measureGroupHeader']) {
+    return measureGroup.items
+        .reduce((result: any,
+                 { measureHeaderItem: { name, format, localIdentifier } }: Execution.IMeasureHeaderItem) => {
+
+            if (result.measuresInSecondAxis.length === secondMeasures.length) {
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: I removed it.

File: src/components/visualizations/chart/legend/helpers.ts [hunk 7]
```diff
@@ -376,3 +377,54 @@ export function getHeatmapLegendConfiguration(
         position: finalPosition
     };
 }
+
+const LEGEND_TEXT = {
+    column: ['Left:', 'Right:'],
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/legend/helpers.ts]: I injected 'intl' to `LegendAxisIndicator`

File: stories/core_components/BarChart.tsx (hunk 2 above)
Comment (naibinh) [reply stories/core_components/BarChart.tsx]: could you point me the issue you see?
you mean column M2 should be displayed after M3?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1921/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1407/ : SUCCESS in 3m 53s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 8]
```diff
@@ -1181,6 +1170,32 @@ function getYAxes(
...
+
+        const { measuresInFirstAxis, measuresInSecondAxis } = assignMeasuresToAxes(secondaryAxisMeasures, measureGroup);
+
+        let firstAxis = createYAxisItem(measuresInFirstAxis, false);
+        let secondAxis = createYAxisItem(measuresInSecondAxis, true);
+
+        if (firstAxis) {
+            firstAxis = {
+                ...firstAxis,
+                percentageFormatKey,
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: is not spread operator '...' missing here?

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 9]
```diff
@@ -1181,6 +1170,32 @@ function getYAxes(
...
+                seriesIndices: range(0, measuresInFirstAxis.length)
+            };
+        }
+        if (secondAxis) {
+            secondAxis = {
+                ...secondAxis,
+                percentageFormatKey,
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: is not spread operator '...' missing here?

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1924/ : SUCCESS in 39s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1410/ : FAILURE in 11m 59s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1928/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1414/ : SUCCESS in 2m 00s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/793/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1932/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/421/ : UNSTABLE in 13m 47s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1934/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1423/ : SUCCESS in 4m 36s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/795/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1935/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/422/ : SUCCESS in 19m 02s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/796/artifact/test-storybook/html-report/index.html

**Comment by naibinh** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/797/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1937/ : SUCCESS in 16s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/424/ : SUCCESS in 20m 15s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1939/ : SUCCESS in 14s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1425/ : FAILURE in 9m 54s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/798/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1940/ : SUCCESS in 14s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/425/ : UNSTABLE in 27m 11s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1941/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1426/ : SUCCESS in 1m 50s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/799/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1942/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/426/ : SUCCESS in 16m 25s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 10]
```diff
@@ -1109,6 +1114,15 @@ function getXAxes(
     }];
 }
 
+function getMeasureFormatKey(measureGroupItems: Execution.IMeasureGroupHeader['measureGroupHeader']['items']) {
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: Why not to use type 'IMeasureHeaderItem'?

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 11]
```diff
@@ -1190,13 +1206,45 @@ function getYAxes(
...
             seriesIndices: range(measureGroupItems.length),
-            ...percentageFormatKey
+            ...getMeasureFormatKey(measureGroupItems)
         }];
     }
 
     return yAxes;
 }
 
+function assignMeasuresToAxes(secondMeasures: string[],
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: you can add return type

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1957/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1439/ : SUCCESS in 2m 32s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1958/ : SUCCESS in 17s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1440/ : SUCCESS in 7m 01s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1984/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1460/ : FAILURE in 6m 12s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/1985/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1461/ : SUCCESS in 1m 54s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/459/ : SUCCESS in 27s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1463/ : SUCCESS in 4m 02s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/808/ : SUCCESS in 16m 04s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/395/ : SUCCESS in 17m 56s

**Pull request merged**